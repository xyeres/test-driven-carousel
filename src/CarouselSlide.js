import PropTypes from 'prop-types';
import React from 'react';
import styled from 'styled-components'

const Img = styled.img`
  object-fit: cover;
  width: 100%;
  height: ${(props) =>
    typeof props.imgHeight === 'number'
      ? `${props.imgHeight}px`
      : props.imgHeight};
`;


const CarouselSlide = ({ imgUrl, imgHeight, description, attribution, ...rest }) => (
  <figure {...rest}>
    <Img src={imgUrl} imgHeight={imgHeight} />
    <figcaption>
      <strong>{description}</strong> {attribution}
    </figcaption>
  </figure>
);

CarouselSlide.propTypes = {
  Img: PropTypes.elementType,
  imgHeight: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  imgUrl: PropTypes.string.isRequired,
  description: PropTypes.node.isRequired,
  attribution: PropTypes.node,
};

CarouselSlide.defaultProps = {
  Img: Img,
  imgHeight: 500,
}

export default CarouselSlide;
