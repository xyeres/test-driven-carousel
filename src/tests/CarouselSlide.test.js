import React from 'react';
import { shallow, mount } from 'enzyme';
import CarouselSlide from '../CarouselSlide';
import styled from 'styled-components';

describe('CarouselSlide', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<CarouselSlide imgUrl="default" description="default" />);
  });

  it('renders a <figure>', () => {
    expect(wrapper.type()).toBe('figure');
  });

  it('renders a props.Img and a <figcaption> as children', () => {
    expect(wrapper.childAt(0).type()).toBe(CarouselSlide.defaultProps.Img);
    expect(wrapper.childAt(1).type()).toBe('figcaption');
  });

  it('passes `imgUrl` through to props.Img', () => {
    const imgUrl = 'https://example.com/image.png';
    wrapper.setProps({ imgUrl });
    const img = wrapper.find(CarouselSlide.defaultProps.Img);
    expect(img.prop('src')).toBe(imgUrl);
  });

  it('uses `description` and `attribution` as the <figcaption>', () => {
    const description = 'A jaw-droppingly spectacular image';
    const attribution = 'Trevor Burnham';
    wrapper.setProps({ description, attribution });
    expect(wrapper.find('figcaption').text()).toBe(
      `${description} ${attribution}`
    );
    expect(wrapper.find('figcaption strong').text()).toBe(description);
  });

  it('passes other props through to the <figure>', () => {
    const style = {};
    const onClick = () => {};
    const className = 'my-car-style';
    wrapper.setProps({ style, onClick, className });
    expect(wrapper.prop('style')).toBe(style);
    expect(wrapper.prop('onClick')).toBe(onClick);
    expect(wrapper.prop('className')).toBe(className);
  });

  it('renders correctly', () => {
    wrapper.setProps({
      description:'descriotn', 
      attribution: 'attrib',
    })
    expect(wrapper).toMatchSnapshot()
  })

  describe('Img', () => {
    let mounted;
    const imgUrl = 'https://example.com/default.jpg';

    beforeEach(() => {
      const Img = CarouselSlide.defaultProps.Img;
      mounted = mount(<Img src={imgUrl} imgHeight={500} />);
    });

    it('renders correctly', () => {
      expect(mounted.find('img')).toMatchSnapshot()
    })

    it('uses imgHeight as the height style property', () => {
      expect(mounted).toHaveStyleRule('height', '500px');
      mounted.setProps({ imgHeight: 'calc(50vh-100px)' });
      expect(mounted).toHaveStyleRule('height', 'calc(50vh-100px)');
    });

    it('allows styles to be overriden', () => {
        const TestImg = styled(CarouselSlide.defaultProps.Img)`
          width: auto;
          height: auto;
          object-fit: fill;
        `;

        mounted = mount(
          <CarouselSlide
            Img={TestImg}
            imgUrl={imgUrl}
            description="This prop is required"
          />
        );

        expect(mounted.find(TestImg)).toHaveStyleRule('width', 'auto');
        expect(mounted.find(TestImg)).toHaveStyleRule('height', 'auto');
        expect(mounted.find(TestImg)).toHaveStyleRule('object-fit', 'fill');
    });
    
  });
});
