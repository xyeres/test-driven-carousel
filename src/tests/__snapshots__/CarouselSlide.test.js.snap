// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CarouselSlide Img renders correctly 1`] = `
.c0 {
  object-fit: contain;
  width: 100%;
  height: 500px;
}

<img
  className="c0"
  src="https://example.com/default.jpg"
/>
`;

exports[`CarouselSlide renders correctly 1`] = `
<figure>
  <CarouselSlide__Img
    alt="an image"
    imgHeight={500}
    src="default"
  />
  <figcaption>
    <strong>
      descriotn
    </strong>
     
    attrib
  </figcaption>
</figure>
`;
